File: solutions/typescript/problems/[0188_Hard] Best Time to Buy and Sell Stock IV.ts
function maxProfit(k: number, prices: number[]): number {
    const n = prices.length;

    // Special case 1: No trades are done
    if (n <= 1 || k <= 0) {
        return 0;
    }

    // Special case 2: More trades allowed than available days - solution for Problem 122
    if (n <= 2 * k) {
        let profit = 0;
        for (let i = 0; i < prices.length - 1; i++) {
            const buyPrice = prices[i];
            const sellPrice = prices[i + 1];

            if (sellPrice > buyPrice) {
                profit += sellPrice - buyPrice;
            }
        }
        return profit;
    }

    /**
     * DP
     * Note: dp[day number][transactions done][0 - not holding stock / 1 - holding stock]
     */
    const dp: number[][][] = [];
    for (let i = 0; i < n; i++) {
        dp[i] = [];
        for (let j = 0; j <= k; j++) {
            dp[i][j] = [];
            // value can be negative when stock is bought and money is spent - need to initialize with -Infinity
            dp[i][j][0] = -Infinity;
            dp[i][j][1] = -Infinity;
        }
    }

    dp[0][0][0] = 0;
    dp[0][1][1] = -prices[0];

    for (let i = 0; i < n; i++) {
        for (let j = 0; j <= k; j++) {
            // Max of two cases: [1] continue not holding any stock; [2] sell stock and get the money
            if (i > 0) {
                dp[i][j][0] = Math.max(dp[i - 1][j][0], dp[i - 1][j][1] + prices[i]);

                // Max of two cases: [1] continue holding stock; [2] buy stock and spend the money
                if (j > 0) {
                    dp[i][j][1] = Math.max(dp[i - 1][j][1], dp[i - 1][j - 1][0] - prices[i]);
                }
            }
        }
    }

    let result = 0;
    for (let j = 0; j <= k; j++) {
        result = Math.max(result, dp[n - 1][j][0]);
    }

    return result;
}

File: solutions/c++/problems/[0188_Hard] Best Time to Buy and Sell Stock IV.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        if (k == 0) return 0;
        if (prices.empty()) return 0;
        int n = prices.size();
        k *= 2;
        vector<vector<int>> dp = vector<vector<int>>(n, vector<int>(k+1, 0));
        int maxP = 0;
        // base cases
        dp[0][k] = 0;
        dp[0][k-1] = -prices[0];
        for (int i = 1; i < n; i++) {
            for (int j = k; j >= k - 1 - i && j >= 0; j--) {
                if (j % 2 == 0) {
                    // not holding stock
                    if (j == k) {
                        dp[i][j] = dp[i-1][j];
                    } else if (j == k - 1 - i) {
                        dp[i][j] = dp[i-1][j+1] + prices[i];
                    } else {
                        dp[i][j] = max(dp[i-1][j], dp[i-1][j+1] + prices[i]);
                    }
                } else {
                    // holding stock
                    if (j == k - 1 - i) {
                        dp[i][j] = dp[i-1][j+1] - prices[i];
                    } else {
                        dp[i][j] = max(dp[i-1][j], dp[i-1][j+1] - prices[i]);
                    }
                }
                maxP = max(maxP, dp[i][j]);
            }
        }
        return maxP;
    }
};